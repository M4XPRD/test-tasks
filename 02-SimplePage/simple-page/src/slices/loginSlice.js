import { createSlice } from '@reduxjs/toolkit';

const loginSlice = createSlice({
  name: 'loginData',
  initialState: {
    login: 'developer21',
    password: '123456',
  },
});

export default loginSlice.reducer;
