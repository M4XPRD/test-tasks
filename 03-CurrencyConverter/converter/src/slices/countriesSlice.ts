import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
import {
  currenciesList, euroData, usdData, audData, nzdData, dkkData,
} from './currenciesData';

export const getCountriesList = createAsyncThunk('countries/getCountriesList', async () => {
  const response = await axios.get('https://restcountries.com/v3.1/all');
  const { data } = response;
  return data;
});

interface Countries {
  countriesList: Country[],
  countriesCurSymbols: {
    [key: string]: string;
  },
  loadingStatus: string,
}

export interface Country {
  name: {
    common: string;
  }
  flag: string;
  currencies: {
    [key: string]: {
      symbol: string;
      name: string;
    };
  };
}

const countriesSlice = createSlice({
  name: 'countries',
  initialState: {
    countriesList: [],
    countriesCurSymbols: {},
    loadingStatus: '',
  } as Countries,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getCountriesList.pending, (state) => {
        state.loadingStatus = 'loading';
      })
      .addCase(getCountriesList.fulfilled, (state, action) => {
        state.loadingStatus = 'finished';
        const countries = action.payload;
        const filteredCountries = countries
          .filter((country: object) => Object.prototype.hasOwnProperty.call(country, 'currencies'))
          .filter((country: object) => {
            const countryTypes = country as Country;
            const [countryCurrency] = Object.keys(countryTypes.currencies);
            if (currenciesList.includes(countryCurrency)) {
              return country;
            }
            return null;
          });

        filteredCountries.push(euroData, usdData, audData, nzdData, dkkData);

        const sortedCountries = filteredCountries
          .sort((a: Country, b: Country) => a.name.common.localeCompare(b.name.common));

        state.countriesList = sortedCountries;
        state.countriesCurSymbols = sortedCountries
          .reduce((acc: { [key: string]: string }, country: Country) => {
            const [currencyShort] = Object.keys(country.currencies);
            acc[currencyShort] = country.currencies[currencyShort].symbol;
            return acc;
          }, {});
      })
      .addCase(getCountriesList.rejected, (state) => {
        state.loadingStatus = 'rejected';
      });
  },
});

export default countriesSlice.reducer;
