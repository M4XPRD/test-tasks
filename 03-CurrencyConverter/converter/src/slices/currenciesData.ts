// Ограничение API, где не все валюты работают и где они дублируются

export const currenciesList = [
  'JPY', 'BGN', 'CZK', 'GBP', 'HUF',
  'PLN', 'RON', 'SEK', 'CHF', 'ISK',
  'NOK', 'HRK', 'RUB', 'TRY', 'BRL',
  'CAD', 'CNY', 'HKD', 'IDR', 'ILS',
  'INR', 'KRW', 'MXN', 'MYR', 'PHP',
  'SGD', 'THB', 'ZAR',
];

export const euroData = {
  name: {
    common: 'European Union',
  },
  flag: '🇪🇺',
  currencies: {
    EUR: {
      name: 'Euro', symbol: '€',
    },
  },
};

export const usdData = {
  name: {
    common: 'United States',
  },
  flag: '🇺🇸',
  currencies: {
    USD: {
      name: 'United States dollar', symbol: '$',
    },
  },
};

export const audData = {
  name: {
    common: 'Australia',
  },
  flag: '🇦🇺',
  currencies: {
    AUD: {
      name: 'Australian dollar', symbol: '$',
    },
  },
};

export const nzdData = {
  name: {
    common: 'New Zealand',
  },
  flag: '🇳🇿',
  currencies: {
    NZD: {
      name: 'New Zealand dollar', symbol: '$',
    },
  },
};

export const dkkData = {
  name: {
    common: 'Denmark',
  },
  flag: '🇩🇰',
  currencies: {
    DKK: {
      name: 'Danish krone', symbol: 'kr',
    },
  },
};
