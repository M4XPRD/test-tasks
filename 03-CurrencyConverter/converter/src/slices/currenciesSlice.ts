import '../navigator.d';
import { createSlice } from '@reduxjs/toolkit';

interface CurrenciesState {
  baseCurrency: string;
  baseCurrencyShort: string;
  exchangeRateURL: string;
  exchangeFrom: string;
  exchangeTo: string;
  currencyAmount: number | string;
  exchangeFromShort: string;
  exchangeToShort: string;
  exchangeAmount: number;
  rates: {
    [key: string]: number;
  },
  axiosError: boolean;
}

interface ExchangePayload {
  newValue: string;
  exchangeOption?: string;
}

interface SetExchangeAction {
  type: string;
  payload: ExchangePayload;
}

const currenciesSlice = createSlice({
  name: 'currencies',
  initialState: {
    baseCurrency: 'USD — United States',
    baseCurrencyShort: 'USD',
    exchangeRateURL: 'https://api.freecurrencyapi.com/v1/latest',
    exchangeFrom: 'USD — United States',
    exchangeFromShort: 'USD',
    exchangeTo: 'RUB — Russia',
    exchangeToShort: 'RUB',
    currencyAmount: '',
    exchangeAmount: 0,
    rates: {},
    axiosError: false,
  } as CurrenciesState,
  reducers: {
    setExchange: (state, action) => {
      const { newValue, exchangeOption } = action.payload;
      const [firstValue, secondValue] = newValue.split(' ');
      const os = navigator.userAgentData.platform;
      if (exchangeOption === 'from') {
        state.exchangeFrom = newValue;
        state.exchangeFromShort = os === 'Windows' ? firstValue : secondValue;
      } else if (exchangeOption === 'to') {
        state.exchangeTo = newValue;
        state.exchangeToShort = os === 'Windows' ? firstValue : secondValue;
      }
    },
    setCurrencyAmount: (state, action) => {
      state.currencyAmount = action.payload;
    },
    setExchangeCurrency: (state, action) => {
      state.exchangeAmount = action.payload;
    },
    setBaseCurrency: (state, action) => {
      const { newValue } = action.payload;
      const [firstValue] = newValue.split(' ');
      state.baseCurrency = newValue;
      state.baseCurrencyShort = firstValue;
    },
    switchCurrencies: (state) => {
      const firstTemp = state.exchangeFrom;
      state.exchangeFrom = state.exchangeTo;
      state.exchangeTo = firstTemp;

      const secondTemp = state.exchangeFromShort;
      state.exchangeFromShort = state.exchangeToShort;
      state.exchangeToShort = secondTemp;
    },
    setRates: (state, action) => {
      const { payload } = action.payload;
      state.rates = payload;
    },
    setRatesError: (state, action) => {
      state.axiosError = action.payload;
    },
  },
});

export const {
  setExchange,
  setCurrencyAmount,
  setExchangeCurrency,
  setBaseCurrency,
  switchCurrencies,
  setRates,
  setRatesError,
} = currenciesSlice.actions;

export type SetExchangeType = (payload: ExchangePayload) => SetExchangeAction;

export default currenciesSlice.reducer;
