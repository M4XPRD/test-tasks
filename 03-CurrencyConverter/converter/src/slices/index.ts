import { configureStore } from '@reduxjs/toolkit';
import countriesReducer from './countriesSlice';
import currenciesReducer from './currenciesSlice';

const store = configureStore({
  reducer: {
    countries: countriesReducer,
    currencies: currenciesReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;
