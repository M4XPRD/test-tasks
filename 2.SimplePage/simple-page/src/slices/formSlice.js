/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit';

const formsSlice = createSlice({
  name: 'form',
  initialState: {
    allChannels: [],
    currentActiveId: null,
  },
  reducers: {
    addChannels: (state, action) => {
      state.allChannels = action.payload;
    },
    addChannel: (state, action) => {
      state.allChannels.push(action.payload);
    },
    removeChannel: (state, action) => {
      const { id } = action.payload;
      const filteredChannels = state.allChannels.filter((channel) => channel.id !== id);
      state.allChannels = filteredChannels;
      state.currentActiveId = state.currentActiveId === id ? 1 : state.currentActiveId;
    },
    renameChannel: (state, action) => {
      const { id, name } = action.payload;
      const updatedChannels = state.allChannels.map((channel) => {
        if (channel.id === id) {
          return { ...channel, name };
        }
        return channel;
      });
      state.currentActiveId = state.currentActiveId === id ? id : state.currentActiveId;
      state.allChannels = updatedChannels;
    },
    setActiveChannel: (state, action) => {
      state.currentActiveId = action.payload;
    },
  },
});

export const {
  addChannels,
  addChannel,
  removeChannel,
  renameChannel,
  setActiveChannel,
} = formsSlice.actions;

export default formsSlice.reducer;
