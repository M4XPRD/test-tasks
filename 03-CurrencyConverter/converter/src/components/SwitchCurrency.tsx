import { Button, Grid } from '@mui/material';
import CompareArrowsIcon from '@mui/icons-material/CompareArrows';
import { useDispatch } from 'react-redux';
import { switchCurrencies } from '../slices/currenciesSlice';

const SwitchCurrency = () => {
  const dispatch = useDispatch();
  return (
    <Grid item xs={12} md="auto">
      <Button
        sx={{
          borderRadius: 1,
          height: '100%',
        }}
        onClick={() => dispatch(switchCurrencies())}
      >
        <CompareArrowsIcon sx={{ fontSize: 30 }} />
      </Button>
    </Grid>
  );
};

export default SwitchCurrency;
