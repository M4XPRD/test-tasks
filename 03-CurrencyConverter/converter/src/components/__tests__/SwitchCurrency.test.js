import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import SwitchCurrency from '../SwitchCurrency';
import store from '../../slices/index';
import { switchCurrencies } from '../../slices/currenciesSlice';

describe('SwitchCurrency component', () => {
  it('Should call dispatch when button is clicked', () => {
    const mockDispatch = jest.fn();
    jest.spyOn(store, 'dispatch').mockImplementation(mockDispatch);

    render(
      <Provider store={store}>
        <SwitchCurrency />
      </Provider>,
    );

    const switchButton = screen.getByTestId('switch-button');
    fireEvent.click(switchButton);

    expect(mockDispatch).toHaveBeenCalledWith(switchCurrencies());
  });
});
