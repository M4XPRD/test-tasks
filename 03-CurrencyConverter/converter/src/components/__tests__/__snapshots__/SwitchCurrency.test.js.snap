// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchCurrency component SwitchCurrency snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-auto css-nmdy5z-MuiGrid-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-udv67l-MuiButtonBase-root-MuiButton-root"
          data-testid="switch-button"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-bq05nv-MuiSvgIcon-root"
            data-testid="CompareArrowsIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M9.01 14H2v2h7.01v3L13 15l-3.99-4v3zm5.98-1v-3H22V8h-7.01V5L11 9l3.99 4z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-auto css-nmdy5z-MuiGrid-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-udv67l-MuiButtonBase-root-MuiButton-root"
        data-testid="switch-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-bq05nv-MuiSvgIcon-root"
          data-testid="CompareArrowsIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9.01 14H2v2h7.01v3L13 15l-3.99-4v3zm5.98-1v-3H22V8h-7.01V5L11 9l3.99 4z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
